core/aggsigdb/memory_v2.go

```diff
@@ -0,0 +1,144 @@
+// Copyright © 2022-2024 Obol Labs Inc. Licensed under the terms of a Business Source License 1.1
+
+package aggsigdb
+
+import (
+	"context"
+	"fmt"
+	"runtime"
+	"sync"
+
+	"github.com/obolnetwork/charon/app/errors"
+	"github.com/obolnetwork/charon/app/log"
+	"github.com/obolnetwork/charon/app/z"
+	"github.com/obolnetwork/charon/core"
+)
+
+// MemDBV2 is a basic memory implementation of core.AggSigDB.
+type MemDBV2 struct {
+	data       sync.Map // map[memDBKey]core.SignedData
+	keysByDuty sync.Map // map[core.Duty][]memDBKey,  Key index by duty for fast deletion.
+	deadliner  core.Deadliner
+	closed     chan struct{}
+}
+
+// NewMemDBV2 creates a basic memory based AggSigDB.
+func NewMemDBV2(deadliner core.Deadliner) *MemDBV2 {
+	return &MemDBV2{
+		// data, keysByDuty are okay to use without explicit initialization
+		deadliner: deadliner,
+		closed:    make(chan struct{}),
+	}
+}
+
+func (m *MemDBV2) store(duty core.Duty, pubKey core.PubKey, data core.SignedData) error {
+	data, err := data.Clone()
+	if err != nil {
+		return err
+	}
+
+	_ = m.deadliner.Add(duty) // TODO(corver): Distinguish between no deadline supported vs already expired.
+
+	key := memDBKey{duty, pubKey}
+
+	if rawExisting, ok := m.data.Load(key); ok {
+		existing, ok := rawExisting.(core.SignedData)
+		if !ok {
+			return errors.New("data stored in aggsigdb not of core.SignedData type", z.Str("key", fmt.Sprintf("%+v", key)))
+		}
+
+		equal, err := dataEqual(existing, data)
+		if err != nil {
+			return err
+		} else if !equal {
+			return errors.New("mismatching data")
+		}
+	} else {
+		m.data.Store(key, data)
```

Comment:
Hmm store was synchronized by channel writing before, so it is true we might have a race between them.

It might be worth to explicitly lock `store()` and have them be always synchronized?

I.e. we're not just syncing access to the data structure, but to the whole logic block.

Reply:
Maybe to lock the whole loop in Store(), to avoid multiple locks.
Also, I think it needs a sync for `data` and `keysByDuty` together? Because these maps have to be consistent.

Reply:
> Maybe to lock the whole loop in Store(), to avoid multiple locks.

Yes, I agree.

> Also, I think it needs a sync for data and keysByDuty together? Because these maps have to be consistent.

We need to lock/unlock at every exported function call I would say.

Reply:
Gave it a shoot: `Store` is locked, so that multiple callers can't potentially overwrite other's duties in the span of their function calls, also made deadline deletion atomic.

`Await` stays as it is, since `Load` should return eagerly.